### ISSUE TITLE
location prev and redirect

### ISSUE THREAD

**Issue by GuillaumeCisco** [faceyspacey/redux-first-router#65]
Hello, I was trying to achieve a common use case where an user wants to access a specific url `/need_to_be_logged` when his session has expired so it will be redirected to `/` with a login page (or a /login view). But after being logged, the user will be redirected to the previous url he wanted to see, not the default HOME . For that we need to keep in memory the url before being redirected to login.

I've thought the `prev` parameter in `location` will be updated after a redirect, example:
```javascript
// this code redirects to LOGIN (which has a login component if the user is no more authenticated)
onBeforeChange: (dispatch, getState, action) => {
        const {user, location} = getState();
        if (action.type !== 'LOGIN' && user && !user.authenticated) {
            dispatch(redirect({type: 'LOGIN'}));
        }
    },
```
So when the user clicks on the button `login` for authenticate himself, it should be redirected to the `need_to_be_logged` view (the previous in memory) and not the default HOME. Of course we would like to use `dispatch(redirect(state.location.prev))`.
So I thought the `state.location.prev` will give me the prev location to redirect to, but this object is empty. Is it normal?
I've seen the prev is linked to the `action.meta.location.prev` so I've tweaked my code as:
```javascript
import {redirect} from 'redux-first-router';
import {clone} from 'lodash';

export default {
    onBeforeChange: (dispatch, getState, action) => {
        const {user, location} = getState();
        const a = clone(action);

        if (action.type !== 'HOME' && user && !user.authenticated && a) {
            const action = redirect({
                type: 'LOGIN',
                meta: {
                    ...a.meta,
                    location: {
                        ...a.meta.location,
                        current: {
                            pathname: '/',
                            type: 'LOGIN',
                            payload: {},
                        },
                        prev: a.meta.location.current,
                    },
                },
            });
            dispatch(action);
        }
    },
};
```
And it works well, but it seems totally wrong to me and like a dirty hack.
Should not the `redirect` function change itself the `prev` attribute?

By the way, I had to clone the action object. Because if I don't, the `action` object becomes `undefined` at some point in the code. Something very very weird is happening on this object, like a mutability by another piece of code.

**Comment by faceyspacey** [comment]
Ultimately, it's not supposed to require you to provide the whole transformed action, and I wouldn't be surprised if it did cause problems.

Your use case is a good one. So basically if you try to enter /private-page, you want to redirect to login and have `prev === /private-page`, correct?

What does it equal now.

I think the problem is that when you submit the login form, it will now link to `/dashboard` and prev will now become `/login`, correct? If that's the case, it's behaving correctly, and what you need to do is instead dispatch an action before redirecting to login that results in the desired page being stored in Redux state. Then after login in, the thunk, do a regular redirect to the saved desired page.

**Comment by GuillaumeCisco** [comment]
Hum, the problem is more about the `redirect` function and the `prev` attribute of the location `reducer`.

1. The user clicks on a link `/private`.
2. The `onBeforeChange` method activate, location reducer is : 
```javascript
asSSR: undefined,
history: undefined,
kind: undefined,
pathname: "/private",
payload: {},
prev: {
    pathname: "",
    payload: {},
    type: "",
},
routesMap: {
     'PRIVATE': {
        pathname : '/private'
    },
    'HOME': {
         pathname : '/'
    },
},
type:"PRIVATE"
```
But be aware that the location logged is from `const {location} = getState();`, not from the redux store.
3. as the user is not authenticated, we want to redirect him to `/` (which will include a login component). So we use the code `dispatch(redirect({type: 'HOME'}))`.
4. When the user clicks on the `login` button, we want to redirect to the `/private` view or the `prev` attribute from location. So we connect our login component to the redux store for getting the `location` reducer. Interestingly, the location reducer is now:
```javascript
asSSR: undefined,
history: undefined,
kind: 'redirect',
pathname: "/",
payload: {},
prev: {
    pathname: "",
    payload: {},
    type: "",
},
routesMap: {
    'PRIVATE': {
        pathname : '/private'
    },
    'HOME': {
        pathname : '/'
    },
},
type:"HOME"
```
I would have think, the `prev` attribute would be:
```javascript
prev: {
    pathname: "/private",
    payload: {},
    type: "PRIVATE",
},
```
Is this normal?
Ultimately, I could myself put the location desired before redirecting to `HOME` in the redux store, and reuse it in my login process.

**Comment by faceyspacey** [comment]
Im pretty sure `/private` will be too far back now. `prev` is now`HOME` when you are redirecting back to `/private`.

Im not sure why it's showing empty--that only happens if there is a page refresh, which for some other reason might be happening. It's what the state is on first load when there is no previous page.

**Comment by GuillaumeCisco** [comment]
What I do not understand, is why after writing `dispatch(redirect({type: 'HOME'}))` in the `onBeforeChange`, I can't see the prev attribute correctly filled in the `onAfterChange`.
If I have a location, make a redirect in the onBeforeChange, should the location in the onAfterChange reflect the change int the prev attribute?

**Comment by faceyspacey** [comment]
it's as if that initial action never happened. I think that's the problem.

**Comment by faceyspacey** [comment]
https://github.com/faceyspacey/redux-first-router/blob/master/src/connectRoutes.js#L279

https://github.com/faceyspacey/redux-first-router/blob/master/src/connectRoutes.js#L301

https://github.com/faceyspacey/redux-first-router/blob/master/src/connectRoutes.js#L254

**Comment by faceyspacey** [comment]
here is where it gets set:

https://github.com/faceyspacey/redux-first-router/blob/master/src/connectRoutes.js#L305

so the prev location only gets set if the action was actually dispatched by the middleware.

Based on the problem you're having, it may be the best idea to set this before short-circuiting. That way when the next action comes through it will retain the `prevLocation` that was redirected from.

What I would do is change this in `node_modules/redux-first-router/dist/connectRoutes.js` and see if you can get it working the way you want.

basically right here:
https://github.com/faceyspacey/redux-first-router/blob/master/src/connectRoutes.js#L254

do something like this:

```js
if (skip) {
  prevLocation = action.meta.location.current
  return
}
```

Give that a try and let me know if it does what you want. Then create a pull request. I'll test it and if it flows with the rest of the system's design, it will be merged.

**Comment by GuillaumeCisco** [comment]
Hey @faceyspacey,
Just made a PR but modified another part of the code, and it works well.
Tell me if it does not break others parts.
https://github.com/faceyspacey/redux-first-router/pull/73

**Comment by GuillaumeCisco** [comment]
Closing issue, merged PR.
